[Lim Zheng Rong] iP

Zeno-Zr opened the pull request:
Requirements completed up to Week 3

Zeno-Zr committed:
add *.class to .gitignore

Zeno-Zr committed:
add skeletal code for greeting and farewell

Zeno-Zr committed:
Add Level-1 requirements and Singlish mode

Program has Level-0 requirements: Add greeting and farewell messages to Duke

Missing Level-1 requirements and customization

Adding the following features allows us to fufill Level-1 requirements
* add greetings during the start of the program
* add farewell messages when user inputs "bye"

Adding customization as encouraged on the CS2113 website
* add a Singlish mode in Duke with variations in text
* add a command "change lang" to toggle it on and off

Zeno-Zr committed:
Add Level-2 requirements

Program has Level-1 requirements and Singlish mode.

Missing Level-2 requirements.

Adding the following features allows us to fufill Level-2 requirements.
* add the ability for user to enter up to 100 strings of text into a
list.
* return the list of texts entered when the command "list" is entered.
* stop program from echoing the text entered.

Zeno-Zr committed:
Update Level-2 requirements

Program has Level-2 requirements.

Missing some features to prevent program from crashing.

Adding the following features allows us to catch these errors.
* add the ability to inform the user that the list is full
* stop the program from adding empty text into the list

Zeno-Zr committed:
Update Level-3 requirements

Program has Level-2 requirements.

Missing level-3 requirements.

Adding the following features allows us to fulfill level-3 requirements.
* add the ability for user to mark and unmark a task on the list
* implement a Task Class as the data structure for tasks in a list

Zeno-Zr committed:
Update codebase to follow a coding standard

Program has Level-3 requirements.

Does not following a particular coding standard.

Implementing the following changes allows us to follow a coding standard
* change codebase to follow the following java coding standard:
https://se-education.org/guides/conventions/java/basic.html

Zeno-Zr committed:
Change how the list of tasks is stored

Tasks are currently stored in a fixed-sized array.

Lacking the ability to store more than 100 tasks in the list.

Adding the following features allows us to add more than 100 tasks to
the list.
* change the underlying data structure to ArrayList
* amend affected methods by the change of data structure

Zeno-Zr committed:
Add support for 3 types of tasks

Program has Level-3 requirements.

Missing level-4 requirements.

Adding the following features partially allows us to fulfill level-4
requirements.
* add the ability for user to label tasks as TODO, DEADLINE or
EVENT when adding them into the list

Zeno-Zr committed:
Add support for 3 types of tasks

Only one type of task is supported (TODO).

Missing the ability to add tasks with dates, which is a level-4
requirement

Adding the following features partially allows us to fulfill the level-4
requirements
* allow user to add 3 types of tasks: "todo", "deadline" and "event"
* allow user to mark and unmark them like previously

Zeno-Zr committed:
Edit runtest.bat file to match Class in folder

Zeno-Zr committed:
Update level-4 requirements

Three type of tasks are supported (TODO, DEADLINE, EVENT).

Failing to catch errors when user enters invalid commands.

Adding the following features allows us to fulfill level-4 requirements
* catch errors when user enters invalid commands
* fix formatting issues such as extra spaces

Zeno-Zr committed:
Improve code quality requirements

Three type of tasks are supported (TODO, DEADLINE, EVENT).

Code violates code quality outlines here:
https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/
codeQuality.

Adding the following features allows us to fulfill code quality
requirements
* refactor methods used in main() in Class Duke
* add Class Greetings for output related text
* shift functions that was previously in main(), such as sayHello() to
Greetings.
* add Javadoc to methods in all the classes
* simplify expressions by refactoring them into a variable
* refactor magic numbers used in into constants

EangJS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+     * @param startDate startDate for EVENT, Deadline date for DEADLINE, null for TODO.
+     * @param endDate endDate for EVENT, null for DEADLINE, null for TODO.
+     */
+    public static void addToList(String line, TypeOfTask typeOfTask, String startDate, String endDate) {
```
Method naming follows java coding standard of camelCase

EangJS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+import java.util.Scanner;
+
 public class Duke {
+    public static final String BYE_COMMAND = "bye";
+    public static final String CHANGE_COMMAND = "change";
+    public static final String LANG_COMMAND = "lang";
+    public static final String LIST_COMMAND = "list";
+    public static final String MARK_COMMAND = "mark";
+    public static final String UNMARK_COMMAND = "unmark";
+    public static final String TODO_COMMAND = "todo";
+    public static final String DEADLINE_COMMAND = "deadline";
+    public static final String DEADLINE_BY = "/by";
+    public static final String EVENT_COMMAND = "event";
+    public static final String EVENT_START_FROM = "/from";
+    public static final String EVENT_END_TO = "/to";
+    /** A fixed sized buffer to trim strings at the right index */
+    public static final int COMMAND_BUFFER = 1;
```
Constant variables are all in capital letters which adheres to the Java Coding standard

EangJS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+            boolean isValidMarkOrUnmarkCommand = (commands[0].equals(MARK_COMMAND) || commands[0].equals(UNMARK_COMMAND)) && commands.length == 2;
+            boolean isTodo = commands[0].equals(TODO_COMMAND);
+            boolean isDeadline = commands[0].equals(DEADLINE_COMMAND);
+            boolean isEvent = commands[0].equals(EVENT_COMMAND);
```
boolean variables have isX which sounds like booleans

EangJS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+            String startOfDate = line.substring(indexOfDate + DEADLINE_BY.length() + COMMAND_BUFFER);
+            // 10 for "deadline " length
+            String descOfTask = line.substring(DEADLINE_COMMAND.length() + COMMAND_BUFFER, indexOfDate);
+            addToList(descOfTask, TypeOfTask.DEADLINE, startOfDate, null);
```
descOfTask is unclear. Maybe use descriptionOfTask instead?

EangJS commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+        this.date = date;
+    }
+
+    /**
+     * Returns the symbol for DEADLINE task
+     * @return D
+     */
+    @Override
```
Correct indentation of comments and follows javadoc requirements

EangJS reviewed (commented):
Hello there! Good Job

adenteo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+import java.util.Scanner;
+
 public class Duke {
+    public static final String BYE_COMMAND = "bye";
+    public static final String CHANGE_COMMAND = "change";
+    public static final String LANG_COMMAND = "lang";
+    public static final String LIST_COMMAND = "list";
+    public static final String MARK_COMMAND = "mark";
+    public static final String UNMARK_COMMAND = "unmark";
+    public static final String TODO_COMMAND = "todo";
+    public static final String DEADLINE_COMMAND = "deadline";
+    public static final String DEADLINE_BY = "/by";
+    public static final String EVENT_COMMAND = "event";
+    public static final String EVENT_START_FROM = "/from";
+    public static final String EVENT_END_TO = "/to";
+    /** A fixed sized buffer to trim strings at the right index */
+    public static final int COMMAND_BUFFER = 1;
+    /** The return result when a substring is not found */
+    public static final int OUT_OF_BOUNDS = -1;
```
Good use of named constants.

adenteo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+        Greeting.sayChangeLanguage(isSinglish);
+        Greeting.sayHello(isSinglish);
+    }
+
```
Very creative and interesting feature here! :)

adenteo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+            }
+        }
+    }
+
```
The code written for the handling of user input is quite clean and easy to understand. Good job!

adenteo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+            // +5 for "/to " length
+            String endOfDate = line.substring(indexOfEndDate + EVENT_END_TO.length() + COMMAND_BUFFER);
+            // 7 for "event " length
+            String descOfTask = line.substring(EVENT_COMMAND.length() + COMMAND_BUFFER, indexOfStartDate);
```
Good use of comments and named constants to explain the use of the substring functions.

adenteo commented on src/main/java/Greeting.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+
+}
```
Very interesting feature! Am looking forward to see how this feature develops.

adenteo reviewed (commented):
Overall, your code was relatively easy to read and follow. Your usage of Javadoc comments made it very easy to understand what every function was for. Your singlish feature is very interesting too! Good job!

Zeno-Zr committed:
Update codebase with suggested improvements

Codebase has level-4 requirements.

improvements can be made to the text printed to output and code quality.

Adding the following features allows us to add improvements to codebase
* add response text when user adds a new task on the list
* edit a variable name (descOfTask to descriptionOfTask) so it is
clearer

Zeno-Zr committed:
Update Singlish text support for all commands

Codebase has level-4 requirements.

Certain commands lack support for Singlish text.

Adding the following features allows us to support Singlish mode
* add singlish text to inform user when task is added
* fix a bug about duplicate greetings when changing language

Zeno-Zr committed:
Update level-5 requirements

Three type of tasks are supported (TODO, DEADLINE, EVENT).

Missing level-5 requirements.

Adding the following features allows us to fulfill level-5 requirements
* add DukeException class for handling custom exceptions
* add exceptions to handle all the possible errors

Adding the following features allows us to improve code quality
* refactor code to make HAPPY path more prominent

Zeno-Zr committed:
Merge branch 'branch-Level-5'

Zeno-Zr committed:
Update level-6 requirements

Program has level-5 requirements.

Missing level-6 requirements.

Adding the following features allows us to fulfill level-6 requirements
* add a way for user to delete a task on the list.

Adding the following features allows us to improve code quality
* add and update javadoc for clarity and spelling mistakes
* update previous implementation of the index value of tasks. Tasks no
longer keep track of their index value. This is replaced by referencing
the index with respect to their order in the list.

Zeno-Zr committed:
Update codebase with level-7 requirements and gitignore

Codebase has level-5 requirements.

Missing level-7 requirements.

Adding the following features allows us to fulfill level-7 requirements
* add persistent storage for list.
* edit icon used for unmark status.

Updating the following for better code management/quality
* add /store/storage-info-txt in gitignore.
* refactor code in Task and its child classes.

Zeno-Zr committed:
add /store/storage-info.txt to gitignore

Zeno-Zr committed:
add /store/storage-info.txt to gitignore

Zeno-Zr committed:
Merge branch 'branch-Level-6'

Zeno-Zr committed:
Merge branch 'branch-Level-7'

Codebase has level-6 requirements.

Missing level-7 requirements.

Updating the following features allows us to fulfill level-7
requirements in the master branch
* add persistent storage for the list of tasks using a text file
* edit icon used for unmark status.
* revamp how the task information is stored in the text file
    * text file now uses a new regex "~;~" instead of a single space to
    avoid conflicts when the description and/or dates contain spaces
    * rearrage the order of arguments stored in the text file and how it
    is processed

Updating the following for better code management/quality
* refactor code in Duke
* refactor code in the child classes of Task
* remove tracking of index in the Task and its child classes since task
order can be changed dynamically

Zeno-Zr committed:
Remove debug lines from Duke class

Codebase has level-7 requirements.

Lingering debug lines in Duke

Updating the following for better code management/quality
* Remove debug lines from Duke
* Add MANIFEST.MF file

maanyos commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+
+    /** Language state of the program. */
+    public static boolean isSinglish = false;
+    /** A fixed sized array to store all the tasks entered from the user. */
```
arraylist is not fixed size

maanyos commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+        Greeting.printHorizontalLines(isSinglish);
+    }
+
+    public static void updateList(Path storageFilePath) throws IOException {
```
consider renaming, can be confused with arraylist of tasks

maanyos commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+        Path storageFilePath = loadStorageFolderAndFile();
+        readAndStoreList(storageFilePath);
+
+        while (true) {
```
use condition here instead of break statement later

maanyos commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+     */
+    private static void addTodo(String line) {
+        // +6 for "deadline " length
+        String desc = line.substring(TODO_COMMAND.length() + COMMAND_BUFFER);
```
use same naming methods as other task types

maanyos reviewed (commented):
good attempt at documentation
please work on the object-orientedness of the code, currently most of the logic is contained in Duke which makes it hard to read and trace through.
try creating new classes for command, parser etc and refactor code accordingly.
refer to addressbook3 if required

Zeno-Zr committed:
Update javadocs and refactor code

Codebase has level-7 requirements.

Outdated javadocs and awkward code blocks.

Updating the following for better code management/quality
* Updating javadocs to reflect current implementation
* Minor refactoring of code

Zeno-Zr committed:
Update codebase to make it more OOP

Codebase has level-7 requirements.

Most of the logic still runs in the main class.

Adding the following features allows us to make the program more OOP
* refactor code in the main class into different classes

Zeno-Zr committed:
Update codebase with level-9 requirements

Codebase has level-7 requirements.

Missing level-9 requirements.

Adding the following features allows us to fulfill level-9 requirements
* adding a way for user to find tasks based on a keyword.

Updating the following for better code management/quality
* renaming method names to avoid ambiguity

Zeno-Zr committed:
Add javadoc to Finder Class

Codebase has level-9 requirements.

Missing javadoc for Finder class.

Updating the following for better code management/quality
* adding javadoc to the Finder class's methods.

Zeno-Zr committed:
Update javadocs for all the classes

Codebase has level-7 requirements.

Javadocs have inconsistent formatting.

Updating the following for better code management/quality
* Update javaDoc formatting accordance to coding standard

Zeno-Zr committed:
Merge pull request #1 from Zeno-Zr/branch-Level-9

Branch level 9

Zeno-Zr committed:
Merge branch 'master' of https://github.com/Zeno-Zr/ip

* 'master' of https://github.com/Zeno-Zr/ip:
  Add javadoc to Finder Class
  Update codebase with level-9 requirements

Zeno-Zr committed:
Merge pull request #2 from Zeno-Zr/heads/A-JavaDoc

Update javadocs for all the classes

Zeno-Zr committed:
Merge branch 'master' of https://github.com/Zeno-Zr/ip

* 'master' of https://github.com/Zeno-Zr/ip:
  Update javadocs for all the classes

Zeno-Zr committed:
v1.0 of user guide

add features and summary table to user guide

Zeno-Zr committed:
Merge pull request #3 from Zeno-Zr/branch-A-UserGuide

v1.0 of user guide

Zeno-Zr committed:
Edit printing formatting in Finder class

Codebase has level-9 requirements.

Missing printing lines for Finder class

Updating the following for consistent formatting
* adding printing lines for Finder class
